tests/test_exchange.py

```diff
@@ -689,67 +689,28 @@ def test_generate_successful_response_on_first_try(normal_exchange):
-
-    def complete(self, model: str, system: str, messages: List[Message], tools: List[Tool]) -> Message:
-        next_item = self.sequence[self.call_count]
-        if self.call_count != len(self.sequence) - 1:
-            self.call_count += 1
-        if isinstance(next_item, Exception):
-            raise HTTPStatusError("Bad Request", request=Mock(), response=Mock())
-        return next_item, Usage(input_tokens=10, output_tokens=10, total_tokens=20)
+def test_rewind_to_last_user_message_in_normal_exchange(normal_exchange):
+    ex = normal_exchange
+    ex.rewind_to_last_user_message()
 
+    assert len(ex.messages) == 0
+    assert len(ex.checkpoint_data.checkpoints) == 0
 
-def test_generate_http_error_recovery_empty_messages():
-    ex = Exchange(
-        provider=MockErrorProvider(
-            sequence=[Message.assistant("Some text"), Exception(), Message.assistant("Some other text")]
-        ),
-        model="gpt-4o-2024-05-13",
-        system="You are a helpful assistant.",
-        moderator=PassiveModerator(),
-    )
-    ex.add(Message.user("Hello"))
-    ex.generate()
-    ex.add(Message.user("Hello again!"))
+    ex.add(Message(role="user", content=[Text("Hello")]))
     ex.generate()
+    ex.add(Message(role="user", content=[Text("Hello")]))
 
-
-def test_generate_http_error_changes_messages():
-    ex = Exchange(
-        provider=MockErrorProvider(
-            sequence=[
-                Message(role="assistant", content=[ToolUse(id="1", name="dishwasher", parameters={})]),
-                Exception(),
-                Message.assistant("I'm done cleaning the dishes."),
-            ]
-        ),
-        model="gpt-4o-2024-05-13",
-        system="You are a helpful assistant.",
-        moderator=PassiveModerator(),
-    )
-    ex.add(Message.user("Hi! Can you clean the dishes for me?"))
-    ex.generate()
-    ex.add(Message(role="user", content=[ToolResult(tool_use_id="1", output="I cleaned the dishes.")]))
-    ex.generate()
+    # testing if it works with a user text message at the end
+    ex.rewind_to_last_user_message()
 
     assert len(ex.messages) == 2
     assert len(ex.checkpoint_data.checkpoints) == 2
-    assert ex.messages[0].text == "Hi! Can you clean the dishes for me?"
-    assert ex.messages[1].text == "I'm done cleaning the dishes."
 
-
-def test_generate_http_error_no_recovery():
-    ex = Exchange(
-        provider=MockErrorProvider(sequence=[Exception()]),
-        model="gpt-4o-2024-05-13",
-        system="You are a helpful assistant.",
-        moderator=PassiveModerator(),
-    )
     ex.add(Message(role="user", content=[Text("Hello")]))
-    with pytest.raises(Exception) as e:
-        ex.generate()
-    assert str(e.value) == "Failed to generate the next message."
+    ex.generate()
+
+    # testing if it works with a non-user text message at the end
+    ex.rewind_to_last_user_message()
+
+    assert len(ex.messages) == 2
+    assert len(ex.checkpoint_data.checkpoints) == 2
```

Comment:
probably worth also adding tool use/results into this test?

Reply:
👍🏼